engine/language-server/src/main/scala/org/enso/languageserver/filemanager/ContentRootManagerActor.scala

```diff
@@ -0,0 +1,176 @@
+package org.enso.languageserver.filemanager
+
+import akka.actor.{Actor, ActorRef, Props}
+import com.typesafe.scalalogging.LazyLogging
+import org.enso.languageserver.data.Config
+import org.enso.languageserver.filemanager.ContentRootManagerActor.ContentRoots
+import org.enso.languageserver.filemanager.ContentRootManagerProtocol._
+import org.enso.languageserver.monitoring.MonitoringProtocol.{Ping, Pong}
+import org.enso.languageserver.util.UnhandledLogging
+import org.enso.polyglot.runtime.Runtime.Api
+
+import java.io.File
+import java.nio.file.{FileSystems, Path => JPath}
+import java.util.UUID
+import scala.jdk.CollectionConverters.IterableHasAsScala
+import scala.util.Try
+
+/** The Actor-based implementation of [[ContentRootManager]] that includes some
+  * additional functionality.
+  *
+  * Aside from providing the desired request-response API, it allows other
+  * Actors to subscribe to content root updates and will send them a
+  * notification whenever a new content root is added.
+  *
+  * It watches the event stream for messages from the RuntimeConnector signaling
+  * that new libraries are loaded, to register their content roots.
+  */
+class ContentRootManagerActor(config: Config)
+    extends Actor
+    with LazyLogging
+    with UnhandledLogging {
+  override def preStart(): Unit = {
+    context.system.eventStream.subscribe(self, classOf[Api.LibraryLoaded])
+  }
+
+  /** @inheritdoc */
+  override def receive: Receive = mainStage(
+    ContentRootManagerActor.initializeRoots(config),
```

Comment:
In the cloud, we have a pool of pre-started language servers. When the user wants to start IDE, we take the server from the pool, mount the user directory, and send `session/initProtocolConnection` to the language server. Until the language server receives this message, it can't access the user directories because they are not mounted.
When the language server receives the `session/initProtocolConnection` it starts initialization. You can subscribe, for example, to the `InitializedEvent.TruffleContextInitialized` and after receiving it, initialize the root directories.

Reply:
After a discussion with @4e6 it turns out we do not need to wait, because the filesystem roots are always present and regarding the project root, which is mounted later, we do not perform any filesystem operations on it here - we just pass what was passed as argument to the LS.